algebird-core/src/main/scala/com/twitter/algebird/maps/prefixsum.scala

```diff
@@ -0,0 +1,224 @@
+ * @tparam K The key type
+ * @tparam V The value type
+ * @tparam P The prefix sum type
+ * @tparam IN The node type of the concrete internal R/B tree subclass
+ * @tparam M The map self-type of the concrete map subclass
+ */
+trait PrefixSumMapLike[K, V, P, IN <: INodePS[K, V, P], M <: PrefixSumMapLike[K, V, P, IN, M] with SortedMap[K, V]]
+  extends NodePS[K, V, P] with OrderedMapLike[K, V, IN, M] {
+
+  /**
+   * A container of all prefix sums over the stored values.  If 'open' is true,
+   * the sums will be for strictly < each key.
+   */
+  def prefixSums(open: Boolean = false) = prefixSumsIterator(open).toIterable
+
+  /**
+   * Iterate over prefix sums for stored values.  If 'open' is true,
+   * the sums will be for strictly < each key.
+   */
+  def prefixSumsIterator(open: Boolean = false) = {
+    val itr = valuesIterator.scanLeft(prefixMonoid.zero)((p, e) => prefixMonoid.inc(p, e))
+    if (open) itr.takeWhile(_ => itr.hasNext) else itr.drop(1)
+  }
+
+  /** equivalent to prefixSum of the right-most key */
+  def sum = this match {
+    case n: INodePS[K, V, P] => n.prefix
+    case _ => prefixMonoid.zero
+  }
+}
+
+sealed trait PrefixSumMap[K, V, P] extends SortedMap[K, V]
+  with PrefixSumMapLike[K, V, P, INodePS[K, V, P], PrefixSumMap[K, V, P]] {
+
+  override def empty = PrefixSumMap.key(keyOrdering).value[V].prefix(prefixMonoid)
+
+  override def toString =
+    "PrefixSumMap(" +
+      iterator.zip(prefixSumsIterator())
+      .map(x => s"${x._1._1} -> (${x._1._2}, ${x._2})").mkString(", ") +
+      ")"
+}
+
+object PrefixSumMap {
+  /**
+   * Instantiate a new empty PrefixSumMap from key, value and prefix types
+   * {{{
+   * import com.twitter.algebird.maps.prefixsum._
+   *
+   * // map strings to integers, using default ordering and standard integer monoid
+   * val map1 = PrefixSumMap.key[String].value[Int].prefix(IncrementingMonoid.fromMonoid[Int])
+   * // Use a custom ordering
+   * val ord: Ordering[String] = ...
+   * val map2 = PrefixSumMap.key(ord).value[Int].prefix(IncrementingMonoid.fromMonoid[Int])
+   * }}}
+   */
+  def key[K](implicit ord: Ordering[K]) = infra.GetValue(ord)
+
+  object infra {
+    /** Mediating class between key method and value method */
+    case class GetValue[K](ord: Ordering[K]) {
+      def value[V] = GetPrefix[K, V](ord)
+    }
+
+    /** Mediating class between value method and prefix method */
+    case class GetPrefix[K, V](ord: Ordering[K]) {
+      def prefix[P](implicit mon: IncrementingMonoid[P, V]): PrefixSumMap[K, V, P] =
+        new Inject[K, V, P](ord, mon) with LNodePS[K, V, P] with PrefixSumMap[K, V, P]
+    }
+  }
+}
+
+/**
+ * A monoid that also supports an 'increment' operation.  This class is intended to encode
+ * semantics similar to Scala's seq.aggregate(z)(seqop,combop), where the standard
+ * Monoid 'plus' corresponds to 'combop' and the 'inc' method corresponds to 'seqop'.
+ */
+trait IncrementingMonoid[T, E] extends Monoid[T] {
```

Comment:
I have some misgivings about the Algebird use of the `prepare` concept.  I believe it doesn't perform well when it is used on monoids with nontrivial structure.

As one example, consider a monoid defined over `Map` where `m1.plus(m1)` (or `m1 ++ m2`) is defined as something such as "merge maps m1 and m2"   Suppose we also want to implement `aggregate` - like behavior over a list of elements.  That is `Map[K, V]`, where `(K, V)` is the element type.  In Scala aggregate, this shows up as `combop(m1, m2)` (the analog of monoid `m1 ++ m2`), and `seqop(m1, (k,v))` (which I will short-hand as operator `m1 + (k,v)`).  Here, `m + (k,v)` is defined as "insert (k,v) into map m" (which is efficient).

Now, compare this to Algebird's aggregators, where instead of defining `+`, the user has to define a `prepare` function, in this case it would be `prepare((k,v)) => Map((k,v))`  That is, for every single pair (k,v) it has to instantiate an entire `Map` object containing a single k/v entry, and then merge it into the result using the monoidal op `++`.

I ran into this when defining the monoid for `TDigest`, where I had to do some gymnastics inside of the `plus` function to make sure that a singleton-tdigest created by `prepare` could be merged in without tripping over quadratic inefficiencies.  A model closer to Scala's `combop` and `seqop` (or `++` and `+`) would be more efficient (and, IMO, actually easier to think about).  My definition of the trait `IncrementingMonoid` was an attempt to capture this more efficent behavior, where I add `inc` (aka `+`) into the trait.